bee-vote/src/fpc.rs

```diff
@@ -0,0 +1,587 @@
+                        opinion_giver,
+                        *selected_count,
+                    ),
+                ));
+            }
+        }
+
+        // Perform all queries.
+        futures::future::join_all(query_futures).await;
+
+        let mut contexts_guard = self.contexts.write().await;
+        let votes_guard = vote_map.read().await;
+
+        // Calculate liked percentage for each vote context.
+        for (object, votes) in votes_guard.iter() {
+            let mut liked_sum = 0.0;
+            let mut voted_count = votes.len() as u32;
+
+            for vote in votes.iter() {
+                match vote {
+                    Opinion::Unknown => voted_count -= 1,
+                    Opinion::Like => liked_sum += 1.0,
+                    _ => {}
+                }
+            }
+
+            // This should never happen – there should always be a context for a given vote.
+            contexts_guard.get_mut(object).unwrap().round_completed();
+
+            // Make sure enough opinions were received for the round to be considered valid.
+            if voted_count < self.min_opinions_received {
+                continue;
+            }
+
+            contexts_guard
+                .get_mut(object)
+                .unwrap()
+                .set_liked(liked_sum / voted_count as f64);
+        }
+
+        // This should never fail – all futures are completed, so only one reference remains.
+        Ok(Arc::try_unwrap(all_queried_opinions).unwrap().into_inner())
+    }
+
+    /// Run a query on a given `OpinionGiver`, to generate opinions on the voting object.
+    async fn do_query(
+        query_ids: &QueryObjects,
+        vote_map: Arc<RwLock<HashMap<VoteObject, Opinions>>>,
+        all_queried_opinions: Arc<RwLock<Vec<QueriedOpinions>>>,
+        opinion_giver: &mut Box<dyn OpinionGiver>,
+        selected_count: u32,
+    ) {
+        // Get opinions from the `OpinionGiver`.
+        let opinions = if let Ok(opinions) = opinion_giver.query(query_ids) {
+            // Sanity check.
+            if opinions.len() != query_ids.conflict_objects.len() + query_ids.timestamp_objects.len() {
+                return;
+            } else {
+                opinions
+            }
+        } else {
+            return;
+        };
+
+        let mut queried_opinions = QueriedOpinions {
+            opinion_giver_id: opinion_giver.id().to_string(),
+            opinions: HashMap::new(),
+            times_counted: selected_count,
+        };
+
+        let mut vote_map_guard = vote_map.write().await;
+
+        // Get opinions on a voting object and add to vote map.
+        let mut query_voting_objects = |objects: &Vec<VoteObject>| {
+            for (i, object) in objects.iter().enumerate() {
+                let mut votes = vote_map_guard
+                    .get(object)
+                    .map_or(Opinions::new(vec![]), |opt| opt.clone());
+
+                for _ in 0..selected_count {
```

Comment:
`selected_count` here is also not immediately clear, after reading this again I think `times_to_count_opinion` is a clearer name.

Reply:
Sure, I can definitely change that.

Reply:
Renamed to `opinion_select_count`, which is a bit briefer but I think still makes it clear that it describes the number of times an opinion gets counted/selected.